aten/src/ATen/native/Distance.cpp

```diff
@@ -52,4 +52,11 @@ Tensor _pdist_backward(const Tensor& grad, const Tensor& self, const double p, c
   return result;
 }
 
+Tensor cosine_similarity(const Tensor& x1, const Tensor& x2, int64_t dim, double eps) {
+  Tensor w12 = at::sum(x1 * x2, dim);
+  Tensor w1 = at::norm(x1, 2, dim);
+  Tensor w2 = at::norm(x2, 2, dim);
+  return w12 / (w1 * w2).clamp_min(eps);
```

Comment:
will it be faster by: `w12.div_( (w1.mul_(w2).clamp_min_(eps) )`?

Reply:
This need to be very careful, because sometimes inplace operators make it impossible to go backward:

```python
import torch

x1 = torch.rand(100, 20, requires_grad=True)
x2 = torch.rand(100, 20, requires_grad=True)

try:
    def cs(x1, x2, dim=1, eps=1e-8):
        w12 = torch.sum(x1 * x2, dim)
        w1 = torch.norm(x1, 2, dim)
        w2 = torch.norm(x2, 2, dim)
        return w12 / (w1 * w2).clamp_min(eps)

    loss = cs(x1, x2).sum()
    loss.backward()
    print('1 works')
except:
    print('1 does not work')


try:
    def cs2(x1, x2, dim=1, eps=1e-8):
        w12 = torch.sum(x1 * x2, dim)
        w1 = torch.norm(x1, 2, dim)
        w2 = torch.norm(x2, 2, dim)
        return w12.div_((w1 * w2).clamp_min(eps))

    loss = cs2(x1, x2).sum()
    loss.backward()
    print('2 works')
except:
    print('2 does not work')


try:
    def cs3(x1, x2, dim=1, eps=1e-8):
        w12 = torch.sum(x1 * x2, dim)
        w1 = torch.norm(x1, 2, dim)
        w2 = torch.norm(x2, 2, dim)
        return w12 / ((w1.mul_(w2)).clamp_min(eps))

    loss = cs3(x1, x2).sum()
    loss.backward()
    print('3 works')
except:
    print('3 does not work')


try:
    def cs4(x1, x2, dim=1, eps=1e-8):
        w12 = torch.sum(x1 * x2, dim)
        w1 = torch.norm(x1, 2, dim)
        w2 = torch.norm(x2, 2, dim)
        return w12 / (w1 * w2).clamp_min_(eps)

    loss = cs4(x1, x2).sum()
    loss.backward()
    print('4 works')
except:
    print('4 does not works')


try:
    def cs5(x1, x2, dim=1, eps=1e-8):
        w12 = torch.sum(x1 * x2, dim)
        w1 = torch.norm(x1, 2, dim)
        w2 = torch.norm(x2, 2, dim)
        return w12.div_((w1.mul_(w2)).clamp_min_(eps))

    loss = cs5(x1, x2).sum()
    loss.backward()
    print('5 works')
except:
    print('5 does not works')


try:
    def cs6(x1, x2, dim=1, eps=1e-8):
        w12 = torch.sum(x1 * x2, dim)
        w1 = torch.norm(x1, 2, dim)
        w2 = torch.norm(x2, 2, dim)
        return w12.div_((w1 * w2).clamp_min_(eps))

    loss = cs6(x1, x2).sum()
    loss.backward()
    print('6 works')
except:
    print('6 does not works')
```

gives

```
1 works
2 works
3 does not work
4 works
5 does not works
6 works
```

Reply:
I changed it to `return w12.div_((w1 * w2).clamp_min_(eps));` instead.